### ISSUE TITLE
Type Error in bng.py

### ISSUE THREAD

**Issue by FFroehlich** [pysb/pysb#287]
I am currently trying to use the PySB functionality in INDRA (see https://github.com/sorgerlab/indra/issues/235). When calling the `export_model('sbml',...)` function I receive the following error:

>Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pysb/bng.py", line 384, in execute
    "\n" +
TypeError: a bytes-like object is required, not 'str'

>During handling of the above exception, another exception occurred:

>Traceback (most recent call last):
  File "testindra.py", line 20, in <module>
    pa.export_model('sbml', file_name='indra.sbml')
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/indra/assemblers/pysb_assembler.py", line 906, in export_model
    exp_str = pysb.export.export(self.model, format)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pysb/export/__init__.py", line 153, in export
    return e.export()
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pysb/export/sbml.py", line 103, in export
    pysb.bng.generate_equations(self.model)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pysb/bng.py", line 524, in generate_equations
    lines = iter(generate_network(model,cleanup=cleanup,verbose=verbose).split('\n'))
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pysb/bng.py", line 499, in generate_network
    bngfile.execute()
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pysb/bng.py", line 387, in execute
    raise BngInterfaceError(e)
pysb.bng.BngInterfaceError: a bytes-like object is required, not 'str'

I am using python 3.5 on macOS 10.12.5. I am using the precompiled BNG 2.2.6 binaries and as far as I can recall I installed PySB via `pip install`.

**Comment by johnbachman** [comment]
I suspect that the immediate error is being masked by the unicode error. The lines in bng.py indicated in the error message are 

```
            if self.verbose:
                for line in iter(p.stdout.readline, b''):
                    print(line, end="")
            (p_out, p_err) = p.communicate()
            if p.returncode:
                raise BngInterfaceError(p_out.rstrip("at line") +
                                           "\n" +
                                           p_err.rstrip())
        except Exception as e:
            raise BngInterfaceError(e)
```

in the line that raises the BngInterfaceError, ```p_out``` will be a bytes-like object, whereas "at line" (the arg to rstrip) and the newline being added to it are Python 3 Unicode strings, which leads to a new error, which is caught by the except block.

So what we're not seeing here is the underlying error being returned by Bng at a particular line of the net file.

Will try to reproduce--thanks!

**Comment by johnbachman** [comment]
@FFroehlich are you still using the same code example from your INDRA issue 235? Or do you have a different example that BNG can compile faster?

**Comment by FFroehlich** [comment]
Currently, yes. I will try to find a less complex setting with which I can reproduce the problem.

**Comment by FFroehlich** [comment]
Okay, I replaced `bp = bel.process_ndex_neighborhood(['BRAF', 'RAF1'])` with `bp = bel.process_ndex_neighborhood(['RPS6KB1'])` and the script finished without throwing an error. So the problem seems to be specific to the produced rule based model.

**Comment by jmuhlich** [comment]
The masking unicode error is fixed in #272 which I will merge soon.

**Comment by johnbachman** [comment]
This took a few minutes to run but it ran successfully for me--image of graph attached. I'm running BNG 2.2.5, however. Will try to reproduce using 2.2.6.
![graph_indra](https://user-images.githubusercontent.com/432756/27387484-db10dba6-5666-11e7-8b86-6f5d3498fbf1.png)

**Comment by FFroehlich** [comment]
For clarification: I could not reproduce the problem with `bp = bel.process_ndex_neighborhood(['RPS6KB1'])` (However, the graph I obtain is the same). It only occurs when using `bp = bel.process_ndex_neighborhood(['BRAF', 'RAF1'])`.

**Comment by johnbachman** [comment]
So as @bgyori pointed out in sorgerlab/indra#235, what is almost certainly happening here is BNG is failing to generate the reaction network for the BRAF-RAF1 model because it contains sets of reactions that yield polymers. This means that the reaction network is infinitely large and BNG will try to enumerate all possible chemical species until it ultimately runs out of memory. This is appearing as a PySB error due to the unicode issues described above, but the underlying issue is the nature of the reaction network. The potential for polymers is indicated by cycles in the contact map (which is generated by Kappa). I've attached the contact map for the BRAF-RAF1 model below:

[cm_braf_raf1.pdf](https://github.com/pysb/pysb/files/1092174/cm_braf_raf1.pdf)

In the contact map the boxes represent the proteins/agents and the yellow circles represent binding sites on those proteins. The edges represent binding interactions between sites. The contact map shows several cycles, for example the path from MAPK1 - CXCR2 - ARRB2 - MAPK1 - ... etc. There is not enough information in the underlying knowledge source (BEL in this case) about the interdependence between sites to assert that CXCR2 can only bind to a MAPK1 that is not already bound (via ARRB2) to another CXCR2, etc. This leads to polymerization and an unbounded reaction network. As Ben mentioned, assembling complexes in a different way in the PySB assembler of INDRA could lead to a different agent/site/rule template, but we haven't implemented a policy that handles this situation yet, in part because we rarely get information from databases or reading that makes use of the state of these complexes (from a causal perspective), so for the time being we tend to filter them out.

The RPS6KB1 model, on the other hand, assembles fine because it contains no cycles, as can be seen in the contact map:

[cm_rps6kb1.pdf](https://github.com/pysb/pysb/files/1092162/cm_rps6kb1.pdf)

...hence the reason you don't get an error. Nevertheless, this model contains over 10 thousand reactions, largely because of the combinatorial nature of the post-translational modification sites on proteins like IRS1 and RPS6KB1. This is why the BNG reaction network generation process takes a long time even for this model.

Going back to BRAF-RAF1, if you modify your script as follows to remove Complexes, it should assemble (though it still takes ~30 min due to the combinatorial explosion due to the large number of PTM sites):

```
import indra
from indra import bel
from indra.statements import *
from indra.assemblers import PysbAssembler
from indra.assemblers import GraphAssembler
from pysb import kappa

pa = PysbAssembler()
ga = GraphAssembler()

bp = bel.process_ndex_neighborhood(['BRAF', 'RAF1'])

stmts = [s for s in bp.statements if not isinstance(s, Complex)]

pa.add_statements(stmts)
ga.add_statements(stmts)

modelpa = pa.make_model()
modelga = ga.make_model()

ga.save_pdf(file_name='graph_indra.pdf')
pa.export_model('sbml', file_name='indra.sbml')
```

Try this out and let me know if you still get an error.

These deceptively simple examples demonstrate why assembling executable models from databases is in general quite tricky. In INDRA we've implemented many tools to turn the "bags of facts" we can get from databases into usable executable models. To get the flavor of these, take a look at the assembly functions exposed in the INDRA assemble_corpus module: http://indra.readthedocs.io/en/latest/modules/tools/index.html#module-indra.tools.assemble_corpus. We're currently working on writing up our approach for publication. In our own applications, we address much of the combinatorial complexity that arises in these models by reducing the model down to the complexes and PTMs that are actually causally relevant to the things (proteins, phosphoforms) that are actually measured in an experimental dataset. A longer discussion on this topic is best reserved for the INDRA repo (sorry to keep bumping back and forth!)

Hope that helps!

**Comment by alubbock** [comment]
This is fixed in my pending PR #272 which is awaiting review.

If you need a quick fix and Python 2 isn't an option, you can make a one line fix to the `subprocess.Popen()` call in `BngFileInterface.execute()` by adding an extra argument `universal_newlines=True`. This will automatically decode `stdout` and `stderr` as `str`, not `bytes`. So it'd look like this:

```python
p = subprocess.Popen(['perl', pf.get_path('bng'),
                      self.bng_filename],
                     cwd=self.base_directory,
                     stdout=subprocess.PIPE,
                     stderr=subprocess.PIPE,
                     universal_newlines=True  # Add this line
                     )
```

**Comment by FFroehlich** [comment]
Okay, adding `stmts = [s for s in bp.statements if not isinstance(s, Complex)]` indeed fixes the problem. Thanks again for the detailed explanation!

**Comment by johnbachman** [comment]
No problem!

**Comment by bgyori** [comment]
I'm reopening this since the fix is not merged yet and causes errors every time I try to build a model in Python 3.

**Comment by johnbachman** [comment]
What underlying BngErrors are you getting that are triggering the unicode issue?

**Comment by bgyori** [comment]
That depends but, for instance, I often get `no reactions were found` in cases where some initial conditions are missing.

**Comment by bgyori** [comment]
Just to clarify, these BNG errors are appearing as expected, the PySB issue is in the error handling code.

**Comment by jmuhlich** [comment]
Fixed by #272.
